tsdb/chunks/chunks.go

```diff
@@ -445,28 +495,52 @@ func (s *Reader) Size() int64 {
 // Chunk returns a chunk from a given reference.
 func (s *Reader) Chunk(ref uint64) (chunkenc.Chunk, error) {
 	var (
-		sgmSeq    = int(ref >> 32)
-		sgmOffset = int((ref << 32) >> 32)
+		// Get the upper 4 bytes.
+		// These contain the segment index.
+		sgmIndex = int(ref >> 32)
+		// Get the lower 4 bytes.
+		// These contain the segment offset where the data for this chunk starts.
+		sgmChunkStart = int((ref << 32) >> 32)
 	)
-	if sgmSeq >= len(s.bs) {
-		return nil, errors.Errorf("reference sequence %d out of range", sgmSeq)
+
+	if sgmIndex >= len(s.bs) {
+		return nil, errors.Errorf("segment index %d out of range", sgmIndex)
 	}
-	chkS := s.bs[sgmSeq]
 
-	if sgmOffset >= chkS.Len() {
-		return nil, errors.Errorf("offset %d beyond data size %d", sgmOffset, chkS.Len())
+	sgmBytes := s.bs[sgmIndex]
+
+	if sgmChunkStart+maxChunkLengthFieldSize > sgmBytes.Len() {
+		return nil, errors.Errorf("segment doesn't include enough bytes to read the chunk size data field - required:%v, available:%v", sgmChunkStart+maxChunkLengthFieldSize, sgmBytes.Len())
 	}
 	// With the minimum chunk length this should never cause us reading
 	// over the end of the slice.
-	chk := chkS.Range(sgmOffset, sgmOffset+binary.MaxVarintLen32)
-
-	chkLen, n := binary.Uvarint(chk)
+	c := sgmBytes.Range(sgmChunkStart, sgmChunkStart+maxChunkLengthFieldSize)
+	chkDataLen, n := binary.Uvarint(c)
 	if n <= 0 {
 		return nil, errors.Errorf("reading chunk length failed with %d", n)
 	}
-	chk = chkS.Range(sgmOffset+n, sgmOffset+n+1+int(chkLen))
 
-	return s.pool.Get(chunkenc.Encoding(chk[0]), chk[1:1+chkLen])
+	chkEncStart := sgmChunkStart + n
+	chkEnd := chkEncStart + chunkEncodingSize + int(chkDataLen) + crc32.Size
+	chkDataStart := chkEncStart + chunkEncodingSize
+	chkDataEnd := chkEnd - crc32.Size
+
+	if chkEnd > sgmBytes.Len() {
+		return nil, errors.Errorf("segment doesn't include enough bytes to read the chunk - required:%v, available:%v", chkEnd, sgmBytes.Len())
+	}
+
+	sum := sgmBytes.Range(chkEnd-crc32.Size, chkEnd)
+	s.crc32.Reset()
+	if _, err := s.crc32.Write(sgmBytes.Range(chkEncStart, chkDataEnd)); err != nil {
+		return nil, err
+	}
+	if act := s.crc32.Sum(s.buf[:0]); !bytes.Equal(act, sum) {
+		return nil, errors.Errorf("unexpected checksum %x, expected %x", act, sum)
+	}
```

Comment:
This might have some effect on performance, but I am not sure why it was not checked before.